controlplane/kubeadm/internal/control_plane.go

```diff
@@ -205,17 +200,124 @@ func (c *ControlPlane) NewMachine(infraRef, bootstrapRef *corev1.ObjectReference
+// It then will recommend the oldest outdated machine in the failure domain with the most outdated machines, if there are any.
+// Lastly it will recommend the oldest machine in the largest failure domain.
+func (c *ControlPlane) NextMachineForScaleDown() *clusterv1.Machine {
+	if c.UnhealthyMachines().Any() {
+		machine := c.UnhealthyMachines().Oldest()
+		c.Logger.Info("Scaling down unhealthy machine", "machine", machine.Name)
+		return machine
+	}
+	if c.MachinesNeedingUpgrade().Any() {
+		machine := c.MachineInFailureDomainWithMostMachines(c.MachinesNeedingUpgrade())
+		c.Logger.Info("Scaling down outdated machine", "machine", machine.Name)
+		return machine
+	}
+	machine := c.MachineInFailureDomainWithMostMachines(c.Machines)
+	c.Logger.Info("Scaling down unneeded machine", "machine", machine.Name)
+	return machine
+}
+
+// ReadyMachines returns machines that are not provisioning or failed.  In the
+// case of 3 node or larger clusters, it excludes unhealthy machines.
+func (c *ControlPlane) ReadyMachines() FilterableMachineCollection {
+	machines := c.Machines.
+		Filter(machinefilters.Not(machinefilters.IsProvisioning)).
+		Filter(machinefilters.Not(machinefilters.IsFailed))
+
+	if c.Machines.Len() < minimumClusterSizeForRemediation {
+		return machines
+	}
+	return machines.Filter(machinefilters.Not(machinefilters.NeedsRemediation))
+}
+
+// NeedsInitialization returns whether the control plane has yet to create the
+// first machine.  Returns false in the case that zero replicas are desired.
+func (c *ControlPlane) NeedsInitialization() bool {
+	return c.Machines.Len() == 0 && *c.KCP.Spec.Replicas > 0
+}
+
+// NeedsScaleDown returns whether the control plane needs to remove a machine
+// from the cluster.  Scale down can be caused by the presence of unhealthy
+// machines or when there are more machines than desired replicas.
+func (c *ControlPlane) NeedsScaleDown() bool {
+	if c.UnhealthyMachines().Any() {
+		c.Logger.Info("Scale down is required because unhealthy machines were found", "machines", c.UnhealthyMachines().Names())
+		return true
+	}
+	if c.Machines.Len() > int(*c.KCP.Spec.Replicas) {
+		c.Logger.Info("Scale down is required", "existing", c.Machines.Len(), "desired", int(*c.KCP.Spec.Replicas))
+		return true
+	}
+	return false
+}
+
+// NeedsScaleUp returns whether the control plane needs to add a machine to the
+// cluster.  Scale up can be caused by an upgrade, where the cluster will
+// replace outdated machines one by one, or by the more common case of having
+// fewer machines than the number of desired replicas.
+func (c *ControlPlane) NeedsScaleUp() bool {
+	if c.NeedsUpgrade() {
+		return true
+	}
+	if c.Machines.Len() < int(*c.KCP.Spec.Replicas) {
+		c.Logger.Info("Scale up is required", "desired", int(*c.KCP.Spec.Replicas), "existing", c.Machines.Len())
+		return true
+	}
+	return false
+}
+
+func (c *ControlPlane) NeedsUpgrade() bool {
+	return c.MachinesNeedingUpgrade().Any() && !c.NeedsScaleDown()
```

Comment:
Tying `NeedsUpgrade()` to `NeedsScaleDown()` seems to be making the code quite difficult to follow/reason through.

Reply:
Yeah I agree the verbiage is confusing. I originally had this change as eliminating `NeedsUpgrade` and reducing the logic to `NeedsScaleUp`/`NeedsScaleDown` but I was afraid people would think this was too radical a change.

Reply:
If it is possible to avoid a separate upgrade workflow and handle everything through scale up/scale down, I think that would be fine, but would need to ensure that upgrade/change rollout prioritization is always maintained over standard scaling operations (which I'm not quite sure we are even doing today).

Reply:
I've changed it so that we only migrate kubeadm secrets etc when doing a rollout (née upgrade). I renamed and made this method private, it's now `needsReplacementMachine`.